src/components/views/Login.tsx

```diff
@@ -123,75 +124,101 @@ class Login extends React.Component<IProps, IState> {
+                            <Form.Group controlId="username">
+                                <Form.Label>
+                                    <FormattedMessage id="login.username" />
+                                </Form.Label>
+                                <Form.Control
+                                    type="text"
+                                    placeholder="Enter username"
+                                    onChange={handleUsrInput}
+                                    value={this.state.username}
+                                    required
+                                />
+                            </Form.Group>
+                            <Form.Group controlId="password">
+                                <Form.Label>
+                                    <FormattedMessage id="login.password" />
+                                </Form.Label>
+                                <Form.Control
+                                    type="password"
+                                    placeholder="Password"
+                                    onChange={handlePwdInput}
+                                    value={this.state.password}
+                                    required
+                                />
+                            </Form.Group>
+                            <Button type="submit" block>
+                                <FormattedMessage id="login.submit" />
+                            </Button>
+                        </Form>
+                    </Card>
+                    {(this.context.serverInfo?.oAuthProviderInfos?.Discord ||
+                        this.context.serverInfo?.oAuthProviderInfos?.GitHub ||
+                        this.context.serverInfo?.oAuthProviderInfos?.Keycloak ||
+                        this.context.serverInfo?.oAuthProviderInfos?.TGForums) && (
+                        <>
+                            <hr />
+                            <Card body>
+                                <Card.Title>OAuth Login</Card.Title>
+                                {Object.keys(this.context.serverInfo.oAuthProviderInfos ?? {}).map(
+                                    provider => {
+                                        const ptheme = providersTheme[provider as OAuthProvider];
+                                        return (
+                                            <Button
+                                                key={provider}
+                                                block
+                                                style={ptheme ? { background: ptheme } : undefined}
+                                                onClick={() =>
+                                                    this.startOAuth(provider as OAuthProvider)
+                                                }>
+                                                {providers[provider as OAuthProvider]}
+                                                &nbsp;
```

Comment:
Can you make use of the d-block classname instead of using whitespace characters

Reply:
Oh that's a space, use span with a ml-1 for the name then, shows how cringe these are

Reply:
```suggestion
```

Yeetus deletus this please